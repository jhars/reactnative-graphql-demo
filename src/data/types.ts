export type ById<T> = { [index: string]: T };

export interface DropPlayerRequestObject {
  leagueId: number;
  playerId: number;
  rosterSpot: RosterSpot;
  rosterId: string;
  playerInfo: Player;
}

export interface SelectDropPlayerObject {
  leagueId: number;
  playerId: number;
  rosterSpot: RosterSpot;
  rosterId: string;
  playerInfo: Player;
}

export enum SortColumnName {
  Name = "Name",
  Points = "Points",
  Position = "Position",
}

export enum Position {
  G = "G",
  D = "D",
  LSM = "LSM",
  SSDM = "SSDM",
  FO = "FO",
  M = "M",
  A = "A",
}

export enum RosterSpot {
  GOALIE = "GOALIE",
  DEFENSE1 = "DEFENSE1",
  DEFENSE2 = "DEFENSE2",
  LSM = "LSM",
  SSDM = "SSDM",
  FO = "FO",
  MIDFIELD1 = "MIDFIELD1",
  MIDFIELD2 = "MIDFIELD2",
  ATTACK1 = "ATTACK1",
  ATTACK2 = "ATTACK2",
  BENCH1 = "BENCH1",
  BENCH2 = "BENCH2",
  BENCH3 = "BENCH3",
  BENCH4 = "BENCH4",
}

interface StatLine {
  points: number;
  goals: number;
  assists: number;
  turnovers: number;
  groundBalls: number;
  faceoffsWon: number;
  saves: number;
}

interface Statistics {
  statLineLastSeason: StatLine;
}

export interface Player {
  id: string;
  firstName: string;
  lastName: string;
  statistics: Statistics;
  position: Position;
}

export interface PlayerData {
  players: Player[];
}

export interface Players {
  data: PlayerData;
}

export interface League {
  id: string;
  title: string;
  teams: Team[];
}

export interface LeaguesData {
  leagues: League[];
}

// interface Leagues {
//   data: LeaguesData;
// }

export interface Team {
  id: string;
  name: string;
  league: League;
}

export interface TeamsData {
  teams: Team[];
}

// interface Teams {
//   data: TeamsData;
// }

export interface Roster {
  id: string;
  teamId: number;
  teamInfo: Team;
  goalie: Player;
  defense1: Player;
  defense2: Player;
  lsm: Player;
  ssdm: Player;
  fo: Player;
  midfield1: Player;
  midfield2: Player;
  attack1: Player;
  attack2: Player;
}

export interface RosterData {
  roster: Roster;
}

interface AddPlayerToTeamRosterData {
  addPlayerToTeamRoster: Roster;
}

export interface AddPlayerToTeamRosterMutationResponse {
  data: AddPlayerToTeamRosterData;
}

//============================
//============================
export interface CurrentUser {
  id: string;
  email: string;
  preferred_username: string;
}
