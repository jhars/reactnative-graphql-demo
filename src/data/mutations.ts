import { gql } from '@apollo/client';

export const ADD_TEAM_TO_LEAGUE = gql`
	mutation AddTeam($name: String!, $leagueId: Int!, $ownerId: ID!) {
	  addTeam(name: $name, leagueId: $leagueId, ownerId: $ownerId) {
	    name
	    leagueId
	    ownerId
	    league {
	      title
	      id
	    }
	  }
	}
`;
export const ADD_PLAYER_TO_TEAM = gql`
mutation AddPlayerToTeamRoster($leagueId: Int!, $teamId: Int!, $playerId: Int!, $position: Position!, $rosterSpot: RosterSpot!,$rosterId: ID!) {
  addPlayerToTeamRoster(leagueId: $leagueId, teamId: $teamId, playerId: $playerId, position: $position, rosterSpot: $rosterSpot, rosterId: $rosterId) {

      goalie {
        lastName
        position
        statistics {
          statLineLastSeason {
            points
          }
        }
      }

      defense1 {
        lastName
        position
        statistics {
          statLineLastSeason {
            points
          }
        }
      }
    
      defense2 {
        lastName
        position
        statistics {
          statLineLastSeason {
            points
          }
        }
      }

      lsm {
        lastName
        position
        statistics {
          statLineLastSeason {
            points
          }
        }
      }

      fo {
        lastName
        position
        statistics {
          statLineLastSeason {
            points
          }
        }
      }

      ssdm {
        lastName
        position
        statistics {
          statLineLastSeason {
            points
          }
        }
      }

      midfield1 {
        lastName
        position
        statistics {
          statLineLastSeason {
            points
          }
        }
      }

      midfield2 {
        lastName
        position
        statistics {
          statLineLastSeason {
            points
          }
        }
      }

      attack1 {
        lastName
        position
        statistics {
          statLineLastSeason {
            points
          }
        }
      }

      attack2 {
        lastName
        position
        statistics {
          statLineLastSeason {
            points
          }
        }
      } 
    }

}
`;