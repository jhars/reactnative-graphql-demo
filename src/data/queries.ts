import { gql } from '@apollo/client';

export const GET_SORTED_PLAYERS = gql`
  query Players($orderBy: PlayerOrder!, $position: Position, $availableForLeagueId: Int) {
    players(orderBy: $orderBy, position: $position, availableForLeagueId: $availableForLeagueId) {
      id
      lastName
      position
      statistics {
        statLineLastSeason {
          points
        }
      }
    }
  }
`;

export const GET_ALL_LEAGUES = gql`
	query Leagues {
	  leagues {
	    id
	    title
	    teams {
	      id
	      name
	    }
	  }
	}
`;

export const GET_TEAMS = gql`
  query Teams($ownerId: ID, $teamsId: Int, $leagueId: Int) {
    teams(ownerId: $ownerId, teamsId: $teamsId, leagueId: $leagueId) {
      id
      name
      league {
        title
        id
      }
    }
  }
`;



export const GET_TEAM_ROSTER = gql`
query Roster($teamId: Int) {
  roster(teamId: $teamId) {

    id
    
    teamInfo {
      name
      id
      league {
        title
        id
      }
    }

    goalie {
      id
      lastName
      position
      statistics {
        statLineLastSeason {
          points
          saves
        }
      }
    }

    defense1 {
    id
    lastName
    position
    statistics {
      statLineLastSeason {
        points
        turnovers
      }
     } 
    }
   
    defense2 {
      id
      lastName
      position
      statistics {
        statLineLastSeason {
          points
          turnovers
      }
     } 
    }

    lsm {
      id
      lastName
      position
      statistics {
        statLineLastSeason {
          points
          groundBalls
        }
      }
    }

    ssdm {
      id
      lastName
      position
      statistics {
        statLineLastSeason {
          points
          groundBalls
        }
      }
    }

    fo {
      id
      lastName
      position
      statistics {
        statLineLastSeason {
          points
          faceoffsWon
        }
      }
    }

    midfield1 {
      id
      lastName
      position
      statistics {
        statLineLastSeason {
          points
          goals
          assists
        }
      }
    }

    midfield2 {
      id
      lastName
      position
      statistics {
        statLineLastSeason {
          points
          goals
          assists
        }
      }
    }

    attack1 {
      id
      lastName
      position
      statistics {
        statLineLastSeason {
          points
          goals
          assists
        }
      }
    }

    attack2 {
      id
      lastName
      position
      statistics {
        statLineLastSeason {
          points
          goals
          assists
        }
      }
    } 
  }
}
`;