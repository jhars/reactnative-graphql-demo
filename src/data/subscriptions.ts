import { gql } from "@apollo/client";

export const TIME_UNTIL_DRAFT = gql`
  subscription DraftTime {
    timeUntilDraft
  }
`;
