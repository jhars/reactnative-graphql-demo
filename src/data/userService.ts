import { fetchAuthSession } from 'aws-amplify/auth';

export const getCurrentUser = async () => {
	console.log("========GET CURRENT USER============");
	try {
		const { accessToken, idToken } = (await fetchAuthSession()).tokens ?? {};
		return {
			email: idToken.payload.email,
			id: idToken.payload.sub,
			preferred_username: idToken.payload.preferred_username
		}
	} catch (err) {
		console.log(err);
	}

} 