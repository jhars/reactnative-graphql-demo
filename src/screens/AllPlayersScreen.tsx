import React, {useState, useEffect} from 'react';
import { View, Text, StyleSheet, FlatList, TouchableOpacity, ActivityIndicator } from 'react-native';
// import {useDispatch, useSelector} from 'react-redux';
// import {fetchPlayersAction} from '../actions/playerAction';
// import {fetchPlayersByStatisticAction} from '../actions/playersOrderedByStatisticAction';
import PlayerRow from '../components/PlayerRow';

//==== GraphQL ========
import { useQuery } from '@apollo/client';
import { GET_SORTED_PLAYERS } from '../data/queries';
import { Players } from '../types';
//=====================

export default function AllPlayersScreen() {


	//pass in players from parent component
	const [ columns, setColumns ] = useState([
    "Position",
    "Name",
    "Points",
  ])

  const [ selectedColumn, setSelectedColumn ] = useState(null)
  const [ pointsSortDirection, setPoinstsSortDirection ] = useState('DESC')
  const [ positionSortDirection, setPositionSortDirection ] = useState('ASC')
  const [ nameSortDirection, setNameSortDirection ] = useState('ASC')



  //==== GraphQL ======== 
	const { loading, error, data, refetch } = useQuery<Players>(GET_SORTED_PLAYERS, {
		variables: {
		  "orderBy": {
		    "field": "points",
		    "order": pointsSortDirection
		  }
		}
	});
	if (loading) return <ActivityIndicator testID="loading" size="large" color="#0000ff" />;
	if (error) return <Text>Error: {error.message}</Text>;
	//=====================
	
  const playerList = data.players;

  //JH-NOTE: Old Redux
	// const playerList = useSelector((state) => state.playersByStat);
	// const dispatch = useDispatch();
	// const searchPlayers = () => {
	//   dispatch(fetchPlayersByStatisticAction({stat: 'points', order: poinstsSortDirection}))
	// }

	// useEffect(()=>{
	//   searchPlayers();
	// }, []);

	const sortTable = (column) => {
	  setSelectedColumn(column)
	  let sortStat
	  let order
	  if (column == "Name") {
	  	sortStat = "lastName"
	  	order = nameSortDirection
	  	setNameSortDirection(order == 'ASC' ? 'DESC' : 'ASC')
	  } else if (column == "Position") {
	  	sortStat = column.toLowerCase()
	  	// order = poinstsSortDirection
	  	order = positionSortDirection
	  	setPositionSortDirection(order == 'ASC' ? 'DESC' : 'ASC')
	  } else if (column == "Points") {
	  	sortStat = column.toLowerCase()
	  	order = pointsSortDirection
	  	// order = positionSortDirection
	  	setPoinstsSortDirection(order == 'ASC' ? 'DESC' : 'ASC')
	  }
	  //JH-NOTE: Old Redux
	  // dispatch(fetchPlayersByStatisticAction({stat: sortStat, order: order}))
	  refetch({"orderBy": {"field": sortStat, "order": order}})
	}


	const tableHeader = () => (
	  <View style={styles.tableHeader}>

		  <TouchableOpacity 
		    style={styles.positionColumnHeader} 
		    onPress={()=> sortTable('Position')}>
		  	<Text style={styles.columnHeaderTxt}>{'Position' + " ▲▼"}</Text>
		  </TouchableOpacity>

		  <TouchableOpacity 
		    style={styles.nameColumnHeader} 
		    onPress={()=> sortTable('Name')}>
		    <Text style={styles.columnHeaderTxt}>{'Name' + " ▲▼"}</Text>
		  </TouchableOpacity>

		  <TouchableOpacity 
		  	style={styles.pointsColumnHeader}
		  	onPress={()=> sortTable('Points')}>
		    <Text style={styles.columnHeaderTxt}>{'Points' + " ▲▼"}</Text>
		  </TouchableOpacity>

	  </View>
	)

	return(
  			<View>
  				<FlatList
  					ListHeaderComponent={tableHeader}
						stickyHeaderIndices={[0]}
				  	data={playerList}
				  	renderItem={({item}) => <PlayerRow indPlayerStats={item} />}
  	      />	
  			</View>
	);
}

const styles = StyleSheet.create({
  container: {
    flex: 1,
    backgroundColor: '#fff',
    alignItems: 'center',
    justifyContent: 'center',
  },
  tableHeader: {
    flexDirection: "row",
    justifyContent: "flex-start",
    alignItems: "center",
    backgroundColor: "#37C2D0",
    height: 50,
    paddingLeft: 15
  },
  tableRow: {
    flexDirection: "row",
    height: 40,
    alignItems:"center",
  },
  positionColumnHeader: {
    alignItems:"left",
  },
  nameColumnHeader: {
    justifyContent: "flex-start",
    alignItems:"center",
    paddingLeft: 20
  },
  pointsColumnHeader: {
  	flex: 2,
  	flexDirection: "row",
  	justifyContent: "flex-end",
  	paddingRight: 60
  },
  columnHeaderTxt: {
    color: "white",
    fontWeight: "bold",
  },
  columnRowTxt: {
    width:"20%",
    textAlign:"center",
  }
});
