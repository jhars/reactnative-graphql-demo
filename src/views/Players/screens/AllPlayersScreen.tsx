import React, { useCallback, useState } from "react";
import {
  ActivityIndicator,
  Alert,
  FlatList,
  Modal,
  Text,
  View,
} from "react-native";
import AddPlayerConfirmationModal from "../components/AddPlayerConfirmationModal";
import AllPlayersTableHeader from "../components/AllPlayersTableHeader";
import PlayerRow from "../components/PlayerRow";
//==== GraphQL ========
import { useQuery } from "@apollo/client";
import { GET_SORTED_PLAYERS } from "../../../data/queries";
import { Player, PlayerData } from "../../../data/types";
//=====================
import { useNavigation, useRoute } from "@react-navigation/native";
import { AvailablePlayersScreenRouteProp } from "../../../navigation/navTypes";

export default function AllPlayersScreen() {
  const navigation = useNavigation();
  const route = useRoute<AvailablePlayersScreenRouteProp>();

  const [addPlayer, setAddPlayer] = useState<Player | null>(null);
  const [modalVisible, setModalVisible] = useState(false);

  const [statCriteria, setStatCriteria] = useState("points");
  const [sortOrder, setSortOrder] = useState("DESC");

  const sortTableCallback = useCallback(
    async (sortStat: string, sortDirection: string) => {
      setStatCriteria(sortStat);
      setSortOrder(sortDirection === "DESC" ? "ASC" : "DESC");
      await handleSortTable();
    },
    [statCriteria],
  );

  const handleSortTable = async () => {
    await refetch({
      orderBy: {
        field: statCriteria,
        order: sortOrder,
      },
    });
  };

  const addPlayerFromPlayerRowCallback = useCallback((player: Player) => {
    selectPlayerToAdd(player);
  }, []);

  const cancelModalCallback = useCallback(() => {
    setModalVisible(false);
  }, []);

  const modalCallBack = useCallback(() => {
    setModalVisible(false);
    navigation.goBack();
  }, []);

  const failedToAddPlayerCallback = useCallback(() => {
    showFailureAlert();
  }, []);

  //==== GraphQL ========
  const { loading, error, data, refetch } = useQuery<PlayerData>(
    GET_SORTED_PLAYERS,
    {
      fetchPolicy: "no-cache",
      variables: {
        orderBy: {
          field: statCriteria,
          order: sortOrder,
        },
        availableForLeagueId:
          Number(route.params?.availableForLeagueId) || null,
        position: route.params?.position || null,
      },
    },
  );

  const showFailureAlert = () => {
    setModalVisible(false);
    Alert.alert(
      "Unable to Add Player",
      "Player may have been added to another team or there may be network issues",
      [
        {
          text: "OK",
          onPress: () => navigation.goBack(),
        },
      ],
    );
  };

  const selectPlayerToAdd = (player: Player) => {
    setAddPlayer(player);
    setModalVisible(true);
  };

  if (loading)
    return <ActivityIndicator testID="loading" size="large" color="#0000ff" />;
  if (error) return <Text>Error: {error.message}</Text>;
  //=====================
  const renderModal =
    addPlayer &&
    route.params?.team &&
    route.params?.rosterId &&
    route.params?.rosterSpot &&
    route.params?.availableForLeagueId;
  return (
    <View style={{ flex: 1 }}>
      {renderModal && (
        <Modal
          animationType="slide"
          transparent={true}
          visible={modalVisible}
          onRequestClose={() => setModalVisible(!modalVisible)}
        >
          <AddPlayerConfirmationModal
            testID={"addPlayerConfirmationModalTestID"}
            lastName={addPlayer.lastName}
            position={addPlayer.position}
            team={route.params.team}
            playerId={Number(addPlayer.id)}
            rosterId={route.params.rosterId}
            rosterSpot={route.params.rosterSpot}
            callback={modalCallBack}
            cancelCallback={cancelModalCallback}
            failedToAddPlayerCallback={failedToAddPlayerCallback}
          />
        </Modal>
      )}

      <FlatList
        contentContainerStyle={{ paddingBottom: 20 }}
        style={{ height: "100vh" }}
        keyExtractor={(item) => item.id}
        ListHeaderComponent={
          <AllPlayersTableHeader
            statCriteria={statCriteria}
            sortOrder={sortOrder}
            callback={sortTableCallback}
            availableForLeagueId={route.params?.availableForLeagueId}
          />
        }
        stickyHeaderIndices={[0]}
        data={data?.players}
        renderItem={({ item }) => (
          <PlayerRow
            player={item}
            addPlayerButton={route.params?.availableForLeagueId ? true : false}
            addPlayerCallback={addPlayerFromPlayerRowCallback}
          />
        )}
      />
    </View>
  );
}
