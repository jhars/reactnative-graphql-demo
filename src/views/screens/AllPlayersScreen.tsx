import React, {useEffect, useState, useCallback} from 'react';
import { Alert, View, Text, FlatList, TouchableOpacity, ActivityIndicator, Modal } from 'react-native';
import PlayerRow from '../components/PlayerRow';
import AddPlayerConfirmationModal from '../components/AddPlayerConfirmationModal';
import AllPlayersTableHeader from '../components/AllPlayersTableHeader';
//==== GraphQL ========
import { useQuery } from '@apollo/client';
import { GET_SORTED_PLAYERS } from '../../data/queries';
import { PlayerData, Team, Position, RosterSpot, SortColumnName, Player} from '../../data/types';
//=====================
import { useNavigation, useRoute } from '@react-navigation/native';
import { AvailablePlayersScreenRouteProp } from '../../navigation/navTypes';
import { ContainerStyles } from '../styles/index';

export default function AllPlayersScreen() {
	const navigation = useNavigation();
	const route = useRoute<AvailablePlayersScreenRouteProp>();

  const [ addPlayer, setAddPlayer ] = useState<Player | null>(null);
  const [ modalVisible, setModalVisible ] = useState(false);
  
  const [ statCriteria, setStatCriteria ] = useState("points");
  const [ sortOrder, setSortOrder ] = useState("DESC");

  const sortTableCallback = useCallback( async (sortStat: string, sortDirection: string) => {
  	setStatCriteria(sortStat)
  	setSortOrder(sortDirection === "DESC" ? "ASC" : "DESC")
  	await handleSortTable()
  }, [statCriteria]);
	  
  const handleSortTable = async () => {
		await refetch({ 
			"orderBy": { 
				"field": statCriteria,
				"order": sortOrder 
			} 
		});	  
  }

  const addPlayerFromPlayerRowCallback = useCallback( (player: Player) => {
  	selectPlayerToAdd(player)
  }, []);

  const cancelModalCallback = useCallback( () => {
  	setModalVisible(false)
  }, []);

  const modalCallBack = useCallback( () => {
  	setModalVisible(false)
  	navigation.goBack();
  }, []);

  const failedToAddPlayerCallback = useCallback( () => {
  	showFailureAlert()
  }, []);

  //==== GraphQL ======== 
	const { loading, error, data, refetch } = useQuery<PlayerData>(GET_SORTED_PLAYERS, {
		fetchPolicy: "no-cache",
		variables: {
		  "orderBy": {
		    "field": statCriteria,
		    "order": sortOrder
		  },
		  "availableForLeagueId": Number(route.params?.availableForLeagueId) || null,
		  "position": route.params?.position || null
		}
	});

	const showFailureAlert = () => {
		setModalVisible(false)
		Alert.alert('Unable to Add Player','Player may have been added to another team or there may be network issues', [
		  {
		    text: 'OK',
		    onPress: () => navigation.goBack(),
		  }
		])
	}	

	const selectPlayerToAdd = (player: Player) => {
		setAddPlayer(player)
		setModalVisible(true)
	}

	if (loading ) return <ActivityIndicator testID="loading" size="large" color="#0000ff" />;
	if (error ) return <Text>Error: {error.message}</Text>;
//=====================
	const renderModal = addPlayer && route.params?.team && route.params?.rosterId && route.params?.rosterSpot && route.params?.availableForLeagueId
	return(
		<View style={{flex: 1}}>
			{ renderModal && 
  			<Modal
  			  animationType="slide"
  			  transparent={true}
  			  visible={modalVisible}
  			  onRequestClose={() => setModalVisible(!modalVisible)}>
	  			  <AddPlayerConfirmationModal 
	  			  	testID={'addPlayerConfirmationModalTestID'}
	  			  	lastName={addPlayer.lastName}
	  			  	position={addPlayer.position}
	  			  	team={route.params.team}
	  			  	playerId={Number(addPlayer.id)}
	  			  	rosterId={route.params.rosterId}
	  			  	rosterSpot={route.params.rosterSpot}
	  			  	callback={modalCallBack}
	  			  	cancelCallback={cancelModalCallback}
	  			  	failedToAddPlayerCallback={failedToAddPlayerCallback}
	  			  />
  			  
			  </Modal>
			}
			
			<FlatList
				contentContainerStyle={{ paddingBottom: 20 }}
				style={{ height: '100vh' }}
				keyExtractor={(item) => item.id}
				ListHeaderComponent={
					<AllPlayersTableHeader 
						statCriteria={statCriteria}
						sortOrder={sortOrder}
						callback={sortTableCallback} 
						availableForLeagueId={route.params?.availableForLeagueId}/>
				}
				stickyHeaderIndices={[0]}
		  	data={data?.players}
		  	renderItem={({item}) =>
	  		  <PlayerRow 
		  				player={item}
		  				addPlayerButton={route.params?.availableForLeagueId ? true : false}
		  				addPlayerCallback={addPlayerFromPlayerRowCallback}
		  		 	/>
		  		}
      />
		</View>
	);
}