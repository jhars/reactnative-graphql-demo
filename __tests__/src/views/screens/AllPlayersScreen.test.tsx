/**
 * @format
 */

import 'react-native';
import React from 'react';
import AllPlayersScreen from '../../../../src/views/screens/AllPlayersScreen';
import {it, describe} from '@jest/globals';
import { MOCK_ALLPLAYERS_RESPONSE } from '../../../mocks/mocks_useQuery_Players'
import { MOCK_QUERY_01 } from '../../../mocks/MOCKS_SHARED'

// JH-NOTE: see RosterRow test for original annoation (must import before 'react-native')
import { render, screen, waitFor } from '@testing-library/react-native';
import {MockedProvider} from '@apollo/react-testing';

import { GET_SORTED_PLAYERS } from '../../../../src/data/queries';

jest.mock("@react-navigation/native", () => {
  const actualNav = jest.requireActual("@react-navigation/native");
  return {
    ...actualNav,
    useNavigation: () => ({
      goBack: jest.fn(),
    }),
    useRoute: () => ({
    	params: {}
    })
  };
});

describe('All Players Screen Test', () => {

	const mocks = [
	  {
	    request: {
	      query: GET_SORTED_PLAYERS,
	      fetchPolicy: "no-cache",
	      variables: { 
	      	orderBy: {
	      		field: 'points',
	      		order: 'DESC'
      		},
      		availableForLeagueId: null,
		  		position: null
	     	}
	    },
	    result: {
	    	data: MOCK_ALLPLAYERS_RESPONSE
	    }
	  }
	];
	
	it('calls the sortTable() function when the Column Header Button is Pressed', async () => {
	  const AllPlayers = render(
	  	<MockedProvider mocks={mocks}>
	  		<AllPlayersScreen testID={"AllPlayersScreenTestID"}/>
	  	</MockedProvider>
	  )
	  
	  const playerName = await screen.findByText('Teat')
	  const testedTree = screen.toJSON()

	  await waitFor(() => {
	  	expect(playerName).toBeVisible()
	    expect(testedTree).toMatchSnapshot()
	  });
	  
	});	

});


