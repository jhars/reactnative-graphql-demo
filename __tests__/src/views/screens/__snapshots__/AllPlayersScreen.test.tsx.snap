// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`All Players Screen Test calls the sortTable() function when the Column Header Button is Pressed 1`] = `
<View>
  <RCTScrollView
    ListHeaderComponent={
      <AllPlayersTableHeader
        callback={[Function]}
        sortOrder="DESC"
        statCriteria="points"
      />
    }
    data={
      [
        {
          "id": "97",
          "lastName": "Teat",
          "position": "A",
          "statistics": {
            "statLineLastSeason": {
              "points": 64,
            },
          },
        },
        {
          "id": "25",
          "lastName": "Nolting",
          "position": "A",
          "statistics": {
            "statLineLastSeason": {
              "points": 39,
            },
          },
        },
        {
          "id": "121",
          "lastName": "Sowers",
          "position": "A",
          "statistics": {
            "statLineLastSeason": {
              "points": 37,
            },
          },
        },
        {
          "id": "131",
          "lastName": "Malone",
          "position": "A",
          "statistics": {
            "statLineLastSeason": {
              "points": 37,
            },
          },
        },
        {
          "id": "95",
          "lastName": "McArdle",
          "position": "A",
          "statistics": {
            "statLineLastSeason": {
              "points": 34,
            },
          },
        },
        {
          "id": "82",
          "lastName": "Fields",
          "position": "A",
          "statistics": {
            "statLineLastSeason": {
              "points": 32,
            },
          },
        },
      ]
    }
    getItem={[Function]}
    getItemCount={[Function]}
    keyExtractor={[Function]}
    onContentSizeChange={[Function]}
    onLayout={[Function]}
    onMomentumScrollBegin={[Function]}
    onMomentumScrollEnd={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    removeClippedSubviews={false}
    renderItem={[Function]}
    scrollEventThrottle={50}
    stickyHeaderIndices={
      [
        0,
      ]
    }
    viewabilityConfigCallbackPairs={[]}
  >
    <View>
      <View
        collapsable={false}
        onLayout={[Function]}
      >
        <View
          style={
            {
              "alignItems": "center",
              "backgroundColor": "rgba(90 130 156 / 1.0)",
              "flexDirection": "row",
              "height": 50,
              "justifyContent": "flex-start",
              "paddingLeft": 15,
            }
          }
        >
          <View
            accessibilityState={
              {
                "busy": undefined,
                "checked": undefined,
                "disabled": undefined,
                "expanded": undefined,
                "selected": undefined,
              }
            }
            accessibilityValue={
              {
                "max": undefined,
                "min": undefined,
                "now": undefined,
                "text": undefined,
              }
            }
            accessible={true}
            collapsable={false}
            focusable={true}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              {
                "alignItems": "center",
                "flex": 1,
                "flexGrow": 2,
                "opacity": 1,
              }
            }
            testID="positionColumnSortHeaderButtonTestID"
          >
            <Text
              style={
                {
                  "color": "white",
                  "fontWeight": "bold",
                }
              }
            >
              Pos.
▲▼
            </Text>
          </View>
          <View
            accessibilityState={
              {
                "busy": undefined,
                "checked": undefined,
                "disabled": undefined,
                "expanded": undefined,
                "selected": undefined,
              }
            }
            accessibilityValue={
              {
                "max": undefined,
                "min": undefined,
                "now": undefined,
                "text": undefined,
              }
            }
            accessible={true}
            collapsable={false}
            focusable={true}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              {
                "flex": 1,
                "flexGrow": 5,
                "opacity": 1,
              }
            }
            testID="nameColumnSortHeaderButtonTestID"
          >
            <Text
              style={
                {
                  "color": "white",
                  "fontWeight": "bold",
                }
              }
            >
              Name
  ▲▼
            </Text>
          </View>
          <View
            accessibilityState={
              {
                "busy": undefined,
                "checked": undefined,
                "disabled": undefined,
                "expanded": undefined,
                "selected": undefined,
              }
            }
            accessibilityValue={
              {
                "max": undefined,
                "min": undefined,
                "now": undefined,
                "text": undefined,
              }
            }
            accessible={true}
            collapsable={false}
            focusable={true}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              {
                "alignItems": "center",
                "flexGrow": 1,
                "opacity": 1,
              }
            }
            testID="pointsColumnSortHeaderButtonTestID"
          >
            <Text
              style={
                {
                  "color": "white",
                  "fontWeight": "bold",
                }
              }
            >
              Points
  ▲▼
            </Text>
          </View>
        </View>
      </View>
      <View
        onFocusCapture={[Function]}
        onLayout={[Function]}
        style={null}
      >
        <View
          style={
            {
              "backgroundColor": "#fff",
              "flex": 1,
              "flexDirection": "row",
              "height": 50,
              "paddingTop": 30,
            }
          }
          testId="playerRowContainerID"
        >
          <View
            style={
              {
                "alignItems": "center",
                "flex": 1,
                "flexGrow": 2,
              }
            }
          >
            <Text>
              A
            </Text>
          </View>
          <View
            style={
              {
                "flex": 1,
                "flexGrow": 5,
              }
            }
          >
            <Text>
              Teat
            </Text>
          </View>
          <View
            style={
              {
                "flexGrow": 1,
              }
            }
          >
            <Text>
              64
            </Text>
          </View>
        </View>
      </View>
      <View
        onFocusCapture={[Function]}
        onLayout={[Function]}
        style={null}
      >
        <View
          style={
            {
              "backgroundColor": "#fff",
              "flex": 1,
              "flexDirection": "row",
              "height": 50,
              "paddingTop": 30,
            }
          }
          testId="playerRowContainerID"
        >
          <View
            style={
              {
                "alignItems": "center",
                "flex": 1,
                "flexGrow": 2,
              }
            }
          >
            <Text>
              A
            </Text>
          </View>
          <View
            style={
              {
                "flex": 1,
                "flexGrow": 5,
              }
            }
          >
            <Text>
              Nolting
            </Text>
          </View>
          <View
            style={
              {
                "flexGrow": 1,
              }
            }
          >
            <Text>
              39
            </Text>
          </View>
        </View>
      </View>
      <View
        onFocusCapture={[Function]}
        onLayout={[Function]}
        style={null}
      >
        <View
          style={
            {
              "backgroundColor": "#fff",
              "flex": 1,
              "flexDirection": "row",
              "height": 50,
              "paddingTop": 30,
            }
          }
          testId="playerRowContainerID"
        >
          <View
            style={
              {
                "alignItems": "center",
                "flex": 1,
                "flexGrow": 2,
              }
            }
          >
            <Text>
              A
            </Text>
          </View>
          <View
            style={
              {
                "flex": 1,
                "flexGrow": 5,
              }
            }
          >
            <Text>
              Sowers
            </Text>
          </View>
          <View
            style={
              {
                "flexGrow": 1,
              }
            }
          >
            <Text>
              37
            </Text>
          </View>
        </View>
      </View>
      <View
        onFocusCapture={[Function]}
        onLayout={[Function]}
        style={null}
      >
        <View
          style={
            {
              "backgroundColor": "#fff",
              "flex": 1,
              "flexDirection": "row",
              "height": 50,
              "paddingTop": 30,
            }
          }
          testId="playerRowContainerID"
        >
          <View
            style={
              {
                "alignItems": "center",
                "flex": 1,
                "flexGrow": 2,
              }
            }
          >
            <Text>
              A
            </Text>
          </View>
          <View
            style={
              {
                "flex": 1,
                "flexGrow": 5,
              }
            }
          >
            <Text>
              Malone
            </Text>
          </View>
          <View
            style={
              {
                "flexGrow": 1,
              }
            }
          >
            <Text>
              37
            </Text>
          </View>
        </View>
      </View>
      <View
        onFocusCapture={[Function]}
        onLayout={[Function]}
        style={null}
      >
        <View
          style={
            {
              "backgroundColor": "#fff",
              "flex": 1,
              "flexDirection": "row",
              "height": 50,
              "paddingTop": 30,
            }
          }
          testId="playerRowContainerID"
        >
          <View
            style={
              {
                "alignItems": "center",
                "flex": 1,
                "flexGrow": 2,
              }
            }
          >
            <Text>
              A
            </Text>
          </View>
          <View
            style={
              {
                "flex": 1,
                "flexGrow": 5,
              }
            }
          >
            <Text>
              McArdle
            </Text>
          </View>
          <View
            style={
              {
                "flexGrow": 1,
              }
            }
          >
            <Text>
              34
            </Text>
          </View>
        </View>
      </View>
      <View
        onFocusCapture={[Function]}
        onLayout={[Function]}
        style={null}
      >
        <View
          style={
            {
              "backgroundColor": "#fff",
              "flex": 1,
              "flexDirection": "row",
              "height": 50,
              "paddingTop": 30,
            }
          }
          testId="playerRowContainerID"
        >
          <View
            style={
              {
                "alignItems": "center",
                "flex": 1,
                "flexGrow": 2,
              }
            }
          >
            <Text>
              A
            </Text>
          </View>
          <View
            style={
              {
                "flex": 1,
                "flexGrow": 5,
              }
            }
          >
            <Text>
              Fields
            </Text>
          </View>
          <View
            style={
              {
                "flexGrow": 1,
              }
            }
          >
            <Text>
              32
            </Text>
          </View>
        </View>
      </View>
    </View>
  </RCTScrollView>
</View>
`;
